PYIC-6088: handle mfa reset request

sam803 opened the pull request:
## Proposed changes

### What changed

- Updates to the Initialise Ipv session handler to check the scope of the received jar and initialise a mfa reset journey if the scope equals "reverification" 
- Additional validation of the journey to the client using the new config parameter `validJourneyTypes`

### Why did it change

When we receive the new JAR from MFA reset, we need to identify this and start an MFA reset journey.

### Issue tracking
https://govukverify.atlassian.net/browse/PYIC-6088

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed

Commit:
handle mfa reset request

remove newline

change error to 403 forbidden

Commit:
Merge branch 'main' into PYIC-6088-handle-authentication-jar

Commit:
update description

Commit:
reduce complexity for sonar cloud

Commit:
sonar cloud didnt like that either - undo

Commit:
refactor to reduce complexity

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -166,7 +172,11 @@ public APIGatewayProxyResponseEvent handleRequest(
-            if (vtr == null || vtr.isEmpty() || vtr.stream().allMatch(String::isEmpty)) {
+
+            String scope = sessionParams.get(REQUEST_SCOPE_KEY);
+            isReverification = "reverification".equals(scope);
```
Could this be a defined constant somewhere?

sam803 replied:
Yes sure - I generally only use constants if its used more than once but not a hill I'd die on :)

Commit:
refactor to reduce complexity

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -175,7 +184,19 @@ public APIGatewayProxyResponseEvent handleRequest(
+            if (error.isPresent()) {
+                LOGGER.error(
+                        LogHelper.buildErrorMessage(
+                                "Validation of the journey type failed.",
```
Could this error message say the something closer to `"User not allowed to initiate this journey type: " + ipvSessionItem.getJourneyType().getType()` maybe.

Maybe the error variable could also be renamed to distinguish it from a generic error. Maybe `forbiddenJourneyError`, I'm not sure though.

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -444,6 +471,14 @@ private void sendInheritedIdentityReceivedAuditEvent(
         }
     }
 
+    private Optional<ErrorResponse> validateJourneyType(String type, String clientId) {
```
It might clarify the method to append "ForClient" to the end of this method name.

Also, if the journeyTypes is null, I think we'd still want to error

Wynndow replied:
I think it might be better to validate that the client is allowed to request the specified scope. This might require some changes to the config that we've set up.
Interestingly the scope can contain more than one value - a space separated string.
The two scopes that we care about are `openid` and `reverification`. We should always receive one of those. I'd suggest we make the validation logic use those, rather than adding a new `type` field to the `IpvJourneyTypes` enum

Commit:
change to boolean

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -175,7 +184,19 @@ public APIGatewayProxyResponseEvent handleRequest(
             IpvSessionItem ipvSessionItem =
-                    ipvSessionService.generateIpvSession(clientOAuthSessionId, null, emailAddress);
+                    ipvSessionService.generateIpvSession(
+                            clientOAuthSessionId, null, emailAddress, isReverification);
```
I'm not sure if it's there's a workaround to get onto a prohibited journey but could make more sense to put validation before this, and use the isReverification for the validation instead

sam803 replied:
hmm yes I think I need to look at a way to do this as currently the generateIpvSession method will store the session even if the journey is invalid, maybe will move the logic to the IPSessionService

sam803 replied:
I've refactored to validate the scope rather than the journey - this way we can validate it before we get here :)

MikeCollingwood reviewed (commented):
This makes sense, the validation of the journey type then generating the ipv session with a different journey type. I like the refactoring too. I did have some Qs, mostly message-based but some on the validation too

Wynndow commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -166,7 +172,10 @@ public APIGatewayProxyResponseEvent handleRequest(
             List<String> vtr = claimsSet.getStringListClaim(REQUEST_VTR_KEY);
-            if (vtr == null || vtr.isEmpty() || vtr.stream().allMatch(String::isEmpty)) {
+
+            String scope = sessionParams.get(REQUEST_SCOPE_KEY);
```
Same as above - we shoould really be using the scope from the claims set, rather than the query param

Wynndow replied:
It would also be good to store the scope on the users clientOauthSessionItem. That way it will be available later in the journey checks.

sam803 replied:
@Wynndow is it definitely going to be in the claims set ? This [PR](https://github.com/govuk-one-login/ipv-stubs/pull/809/files) only adds it to the request, so might need to add it there as well

Wynndow replied:
Yep - the claims set will always contain it. I think @thebauSoftwire might even be adding validation that it does.
And yep, it'll need adding to the stub. I'm working on an orch stub PR at the moment so can add it.

sam803 replied:
thanks, if you haven't already done it theres this pr https://github.com/govuk-one-login/ipv-stubs/pull/814

Wynndow replied:
No, all good, approved

Wynndow commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -156,6 +160,8 @@ public APIGatewayProxyResponseEvent handleRequest(
                             JWEObject.parse(sessionParams.get(REQUEST_PARAM_KEY)),
                             configService.getSsmParameter(JAR_KMS_ENCRYPTION_KEY_ID));
+
+            String clientId = sessionParams.get(CLIENT_ID_PARAM_KEY);
```
We get the client_id as a query param, but it will also be coming in the claims set inside the JAR. When we have the value in the JAR we should be using it from there, as then we have the security that it's signed by the client.

Wynndow commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/config/ConfigurationVariable.java:
```diff
@@ -10,6 +10,7 @@ public enum ConfigurationVariable {
     CI_CONFIG("self/ci-config"),
     CI_SCORING_THRESHOLD("self/ciScoringThreshold"),
     CLIENT_ISSUER("clients/%s/issuer"),
+    CLIENT_VALID_JOURNEY_TYPES("clients/%s/validJourneyTypes"),
```
As above - I think maybe we should change this to `validScopes`

Wynndow commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java:
```diff
@@ -98,7 +101,11 @@ public IpvSessionItem generateIpvSession(
+                ipvSessionItem.setJourneyType(IpvJourneyTypes.MFA_RESET);
+            } else {
+                ipvSessionItem.setJourneyType(IpvJourneyTypes.INITIAL_JOURNEY_SELECTION);
+            }
```
```suggestion
            ipvSessionItem.setJourneyType(isReverification ? IpvJourneyTypes.MFA_RESET : IpvJourneyTypes.INITIAL_JOURNEY_SELECTION)
```

Commit:
refactor to validate the scope rather than the journey

Commit:
rebase

Commit:
remove uneeded line

Commit:
use isEmpty

Commit:
save scope in clientOAuthSessionItem

MikeCollingwood commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -87,6 +89,46 @@ public JWTClaimsSet validateRequestJwt(SignedJWT signedJWT, String clientId)
+                        OAuth2Error.ACCESS_DENIED.setDescription(
+                                "Client not allowed to issue a request with this scope"));
+            }
+            LogHelper.attachClientIdToLogs(clientId);
```
Very minor thoughts:
- Could be good to attach client id at the start so don't need to manually add them in
- Log messages could have commonly capitalised first letter

MikeCollingwood reviewed (dismissed):
This looks good and makes sense!

Wynndow commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -87,6 +89,46 @@ public JWTClaimsSet validateRequestJwt(SignedJWT signedJWT, String clientId)
+        try {
+            String scope = claimsSet.getStringClaim("scope");
+            var validScopes = configService.getValidClientScopes(clientId);
+            if (validScopes.isEmpty() || !validScopes.contains(scope)) {
```
I think there's another case here worth checking. We would never expect to receive `openid` and `reverification` in the scope - it can be a space separated (I think) list of scopes.

Wynndow committed:
PYIC-6088: Refactor to handle scopes

This introduces the use of the nimbus Scope object. Scopes can be a
space separated list of values. And that is in fact what we already
receive from the orchestrator. Using the object will allow us to to more
advanced things in the future.

This also adds a case when validating the scope in the JAR request. We
should never recieve both openid and reverification in the same request.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/mfa-reset.yaml:
Do you think this is going to be specific to MFA reset, or a general 'reverification' journey?

Wynndow replied:
Better to keep it generic if we think we might reuse it - i'll change it

Wynndow committed:
PYIC-6088: Move scope validation into claimsSet validation

This pushes the specific validation of the scope claim further into the
JarValidator. This allows us to throw an exception and catch it to be
rethrown as a recoverable exception in line with how we do it elsewhere.

This also adds scope as a required claim to the Jar validator.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -87,6 +94,62 @@ public JWTClaimsSet validateRequestJwt(SignedJWT signedJWT, String clientId)
+        var validClientScopes =
+                Scope.parse(configService.getSsmParameter(CLIENT_VALID_SCOPES, clientId));
+        return !requestedScopes.stream()
+                .allMatch(scope -> validClientScopes.contains(scope.getValue()));
```
Fine to leave as-is, but as a minor style thing, I generally try to keep methods 'positive' and include a verb - e.g. `hasRequiredScope` and `hasValidScopesForClient`.

Wynndow replied:
No, that's much better. My brain sometimes has a meltdown when dealing with double negatives so will fix it up

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/validation/JarValidator.java:
```diff
@@ -29,19 +30,25 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final String REDIRECT_URI_CLAIM = "redirect_uri";
-    public static final String CLAIMS_CLAIM = "claims";
+    private static final List<String> ONE_OF_REQUIRED_SCOPES = List.of("openid", "reverification");
```
(Should these scopes be constants somewhere - `reverification` is also used in `InitialiseIpvSessionHandler`)

Wynndow replied:
Yeah, that's a good idea

Wynndow committed:
PYIC-6088: Use Instant when validating JAR ttl

Instant.now() will return a time in UTC. We were using LocalDateTime and
OffsetDateTime which are timezone unaware, so will differ depending on
the system clock that is generating them.

This was manifesting with tests passing locally but not in GHA. Local
tests were getting a BST time, and GHA were getting a UTC time.

Wynndow committed:
PYIC-6088: Rename MFA reset journey to reverification

Calling it reverification keeps it generic and we might use it for
something other than MFA reset in future.

Wynndow committed:
PYIC-6088: Rename scope helper methods

To make them easier to grok

Wynndow committed:
PYIC-6088: Create ScopeConstants

This creates a utility class to store constants related to scopes
